root/programs/MultiPurposeAuthSite/MultiPurposeAuthSite/Controllers/AccountController.cs

```diff
@@ -0,0 +1,2221 @@
+                                access_token, access_secret,
+                                ASPNETIdentityConfig.TwitterAuthenticationClientId,
+                                ASPNETIdentityConfig.TwitterAuthenticationClientSecret);
+
+                            email = (string)myInfo["email"]; // Microsoft.Owin.Security.Twitterでは、emailClaimとして取得できない。
+                            emailClaim = new Claim(ClaimTypes.Email, email); // emailClaimとして生成
+                        }
+                    }
+                    else
+                    {
+                        // emailClaimが取得できた場合、
+                        email = emailClaim.Value;
+                    }
+                    #endregion
+
+                    string uid = "";
+                    if (ASPNETIdentityConfig.RequireUniqueEmail)
+                    {
+                        uid = email;
+                    }
+                    else
+                    {
+                        uid = name;
+                    }
+
+                    if (!string.IsNullOrEmpty(email)
+                        && !string.IsNullOrEmpty(name))
+                    {
+                        // クレーム情報（e-mail, name情報）を取得できた。
+
+                        // 既存の外部ログインを確認する。
+                        ApplicationUser user = await UserManager.FindAsync(login);
+
+                        if (user != null)
+                        {
+                            // 既存の外部ログインがある場合。
+
+                            // ユーザーが既に外部ログインしている場合は、クレームをRemove, Addで更新し、
+                            result = await UserManager.RemoveClaimAsync(user.Id, emailClaim); // del-ins
+                            result = await UserManager.AddClaimAsync(user.Id, emailClaim);
+                            result = await UserManager.RemoveClaimAsync(user.Id, nameClaim); // del-ins
+                            result = await UserManager.AddClaimAsync(user.Id, nameClaim);
+
+                            // SignInAsyncより、ExternalSignInAsyncが適切。
+
+                            //// 通常のサインイン
+                            //await SignInManager.SignInAsync(
+
+                            // 既存の外部ログイン・プロバイダでサインイン
+                            signInStatus = await SignInManager.ExternalSignInAsync(
+                                                 loginInfo: externalLoginInfo,
+                                                 isPersistent: false); // 外部ログインの Cookie 永続化は常に false.
+
+                            // AppScan指摘の反映
+                            this.FxSessionAbandon();
+                            // SessionIDの切換にはこのコードが必要である模様。
+                            // https://support.microsoft.com/ja-jp/help/899918/how-and-why-session-ids-are-reused-in-asp-net
+                            Response.Cookies.Add(new HttpCookie("ASP.NET_SessionId", ""));
+
+                            // イベント・ログ出力
+                            Logging.MyOperationTrace(string.Format("{0}({1}) did ext sign in.", user.Id, user.UserName));
```

Comment:
@daisukenishino 

did ext sign in -> has signed in with a verified external account